Event 24h notification

pedrocnogueira opened the pull request:
Adiciona cronograma para enviar uma notificação para o usuário 24 horas antes da sessão do evento me que ele está inscrito.

Adiciona a funcionalidade de "shouldCancelWithConfirmation" na página de ticket. Modal de confirmação nos prints abaixo, tanto em mobile quanto desktop.

<img width="1511" height="824" alt="image" src="https://github.com/user-attachments/assets/d4406dc0-c34f-4dde-8bca-eb9a3e2d9396" />
<img width="380" height="670" alt="image" src="https://github.com/user-attachments/assets/770eb930-0888-441b-b297-85ca7efd6011" />


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Upcoming event notifications for subscribers via push and WhatsApp (sent daily at 16:00 UTC).
  - Optional confirmation dialog before canceling a ticket.

- **Enhancements**
  - Cancellation flow now preserves your latest pre‑screening answers.
  - Notification templates, channels, base URL and backoffice credentials configurable at runtime.
  - Notifications tracked to avoid duplicate sends.
  - Improved Brazil timezone handling for scheduling.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

pedrocnogueira committed:
Add upcoming event notifications feature

- Introduced a new cron job to send notifications for upcoming events every 15 minutes.
- Added functions to retrieve events eligible for notifications and to mark notifications as sent.
- Implemented event handler for sending notifications via OneSignal and WhatsApp.
- Updated event schema to track when notifications are sent.
- Enhanced ticket cancellation process with a confirmation dialog in the UI.

pedrocnogueira committed:
Refactor upcoming notifications event handler

- Removed the redundant 'data' field and replaced it with 'custom_data' in the notification payload for OneSignal.
- Streamlined the structure of the notification body for improved clarity and functionality.

guim4dev commented on server/inngest/crons/upcomingEventNotifications.ts:
```diff
@@ -0,0 +1,30 @@
+  async () => {
+    const events = await getEventsToSendUpcomingNotifications();
+    for (const event of events) {
+      await (inngest.send as any)({
```
no puede esse any

guim4dev commented on server/inngest/crons/upcomingEventNotifications.ts:
```diff
@@ -0,0 +1,30 @@
+    id: "find-events-to-send-upcoming-notifications",
+  },
+  {
+    cron: "*/15 * * * *", // every 15 minutes
```
uma vez ao dia (quando enviamos as notifs)

pedrocnogueira committed:
Update upcoming event notifications and enhance ticket cancellation UI

- Changed the cron job schedule to send notifications daily at 2 PM instead of every 15 minutes.
- Modified the logic for retrieving upcoming events to focus on the next day's events.
- Added a check for missing Hemocione ID credentials in the notification handler.
- Improved the ticket cancellation process by refining the confirmation dialog in the UI.

pedrocnogueira committed:
Refactor upcoming event notifications to use subscriptions

- Replaced event retrieval logic with subscription-based notifications for upcoming events.
- Grouped subscriptions by eventSlug for efficient processing and notification sending.
- Enhanced notification handler to filter subscriptions that require notifications based on their schedule.
- Added a new field in the subscription model to track when notifications are sent.

guim4dev commented on server/inngest/crons/upcomingEventNotifications.ts:
```diff
@@ -0,0 +1,43 @@
+    id: "find-events-to-send-upcoming-notifications",
+  },
+  {
+    cron: "0 14 * * *", // every day at 2 PM
```
this is 2 PM in UTC, which means 11 AM in BRT. it this the behavior you want?

guim4dev commented on server/services/subscription.ts:
```diff
@@ -145,3 +160,40 @@ export async function deleteSubscription(
+  tomorrow.setHours(0, 0, 0, 0); // Start of tomorrow
+  
+  const endOfTomorrow = new Date(tomorrow);
+  endOfTomorrow.setHours(23, 59, 59, 999); // End of tomorrow
```
this whole logic is being done in utc (the day in UTC will have a shift of 3 hours from our brazilian time). is this the behavior you want?

guim4dev commented on server/inngest/eventHandlers/sendUpcomingNotifications.ts:
```diff
@@ -0,0 +1,125 @@
+    await Promise.all(
+      subscriptionsToNotify.map(sub => 
+        markSubscriptionUpcomingNotificationAsSent(String(sub._id))
+      )
```
this will trigger N calls to the database that do the same thing (N network calls and etc - being N the number of subscriptions). Can we do that with a single operation? using a updateMany call

guim4dev reviewed (changes requested):
i'd rather we do this based off on the event just to make sure we dont overload calls at the same time by processing a lot of subscriptions from different events

pedrocnogueira committed:
refactor: optimize notification system and fix Brazil timezone handling

- Change cron to process events instead of all subscriptions at once
- Implement batch processing (100 users per batch) to avoid simultaneous API calls
- Use getEventsToSendUpcomingNotifications() instead of getSubscriptionsToSendUpcomingNotifications()
- Add detailed logging for batch processing and monitoring

- Fix timezone handling to use Brazil timezone (UTC-3) consistently
- Update cron schedule from 14:00 UTC to 16:00 UTC (13:00 Brazil time)
- Create brazilTimezone utility functions for standardized date calculations
- Refactor all date operations to use Brazil timezone instead of UTC

This should prevent overloading the notification API when multiple events have many subscribers and ensure notifications are sent at the correct Brazil time. Is this the approach you were looking for?

pedrocnogueira committed:
Increase batch size for processing upcoming notifications from 100 to 200

pedrocnogueira committed:
perf: optimize database operations for notification marking

- Replace individual findByIdAndUpdate calls with single updateMany operation
- Add markMultipleSubscriptionsUpcomingNotificationAsSent() function
- Reduce N database calls to 1 for marking subscriptions as notified
- Improve performance and reduce network overhead for large subscription batches

Before: 500 subscriptions = 500 DB calls
After: 500 subscriptions = 1 DB call

pedrocnogueira committed:
fix: enhance validation for notification credentials

- Added checks for missing OneSignal and WhatsApp template IDs in the notification handler.
- Updated error message to specify which credentials are missing, improving clarity for debugging.

pedrocnogueira commented on server/inngest/eventHandlers/sendUpcomingNotifications.ts:
```diff
@@ -0,0 +1,125 @@
+    endOfTomorrow.setHours(23, 59, 59, 999);
+    
+    const subscriptionsToNotify = subscriptions.filter(sub => 
+      sub.schedule.startAt >= tomorrow && 
+      sub.schedule.startAt <= endOfTomorrow &&
+      !sub.notificationsUpcomingSentAt
+    );
+    
+    if (!subscriptionsToNotify.length) return { skipped: true, reason: "No subscriptions need notifications" };
+    
+    const userIds = subscriptionsToNotify.map(s => s.hemocioneId).filter(Boolean) as string[];
+
+    const config = useRuntimeConfig();
+    const hemocioneIdBaseUrl = config.hemocioneIdBaseUrl;
+    const backofficeToken = config.hemocioneIdBackofficeToken;
+    const oneSignalTemplateId = config.oneSignalTemplateId;
+    const whatsappTemplateName = config.whatsappTemplateName;
+
+    if (!backofficeToken || !hemocioneIdBaseUrl) {
+      return { skipped: true, reason: "Missing Hemocione ID credentials/config" };
+    }
```
Corrigido no commit fix: enhance validation for notification credentials

pedrocnogueira committed:
Update batch size for processing upcoming notifications from 50 to 200 users

pedrocnogueira commented on server/inngest/crons/upcomingEventNotifications.ts:
```diff
@@ -0,0 +1,31 @@
+    cron: "0 16 * * *", // every day at 1 PM Brazil time (16:00 UTC)
+  },
+  async () => {
+    const events = await getEventsToSendUpcomingNotifications();
+    
+    // Send one notification event per event (the handler will process subscriptions individually)
+    for (const event of events) {
+      await inngest.send({
+        name: sendUpcomingEventName,
+        data: {
+          slug: event.slug,
+        },
+      });
+    }
```
@guim4dev Isso aqui tá levando pro lado contrário do que você tinha falado de ir pelos eventos. Como prosseguir?

pedrocnogueira committed:
refactor: simplify upcoming notifications logic and remove unused timezone functions

- Removed unnecessary Brazil timezone functions from event and subscription services.
- Updated notification filtering logic to focus on subscriptions that haven't been notified yet.
- Enhanced clarity in the upcoming notifications handler by streamlining date range checks.

pedrocnogueira committed:
refactor: improve notification handling and error management in upcoming notifications

- Enhanced the upcoming notifications handler to track successfully sent subscription IDs, allowing for more accurate notification marking.
- Implemented error handling to ensure failed batches do not affect the overall notification process.
- Updated the database operation to only mark successfully sent notifications, improving reliability and clarity in the notification process.
- Added the `_id` field to the subscription selection to facilitate tracking of sent notifications.

pedrocnogueira commented on server/inngest/eventHandlers/sendUpcomingNotifications.ts:
```diff
@@ -0,0 +1,136 @@
+    if (!subscriptions.length) return { skipped: true, reason: "No subscribers" };
+    
+    // Filter subscriptions that need notifications (schedule.startAt is tomorrow and not yet notified)
+    const tomorrowStart = getBrazilTomorrowStart();
+    const tomorrowEnd = getBrazilTomorrowEnd();
+    
+    const subscriptionsToNotify = subscriptions.filter(sub => 
+      sub.schedule.startAt >= tomorrowStart && 
+      sub.schedule.startAt <= tomorrowEnd &&
+      !sub.notificationsUpcomingSentAt
+    );
```
fixed no commit: refactor: improve notification handling and error management

pedrocnogueira commented on server/inngest/eventHandlers/sendUpcomingNotifications.ts:
```diff
@@ -0,0 +1,136 @@
+    const BATCH_SIZE = 200; // Process up to 200 users at a time
+    let processedCount = 0;
+    
+    for (let i = 0; i < subscriptionsToNotify.length; i += BATCH_SIZE) {
```
fixed no commit: refactor: improve notification handling and error management

pedrocnogueira committed:
refactor: streamline upcoming notifications processing and enhance error handling

- Simplified the notification sending logic by removing batch processing and directly mapping subscriptions for notifications.
- Improved error handling to ensure all subscriptions are marked as notified, regardless of individual send failures.
- Updated the notification payload structure for better clarity and consistency in data handling.

guim4dev commented:
## 🤖 Review automático — PR #32: Event 24h notification

**Status:** ✅ CI passa | ✅ Sem conflitos | 🔴 Changes Requested (guim4dev)

**Situação:**
- Mergeável: sim (sem conflitos de merge)
- MergeState: BLOCKED — há "Changes Requested" aberto pelo @guim4dev
- Múltiplos issues críticos apontados pelo CodeRabbit ainda pendentes:
  1. **WhatsApp template parameters** estão aninhados incorretamente (array-of-arrays) — deveria enviar por usuário com parâmetros flat
  2. **Janela de 24h**: o filtro usa "amanhã" no timezone BR, não uma janela rolling de now+24h — pode notificar fora do tempo correto dependendo do horário do cron
  3. **Variável  não utilizada** (TypeScript/ESLint error)
  4. **Formatting (prettier)** — muitos erros pendentes nos arquivos modificados
  5. **markSubscriptions** deve ser feito apenas para envios bem-sucedidos

**Ação recomendada:** O author precisa endereçar os pontos levantados pelo @guim4dev e pelo CodeRabbit antes do merge. Prioridade nos itens 1 (WhatsApp payload crítico) e 2 (timing da janela de notificação).

Commit:
fix: WhatsApp payload per-user, rolling 24h window, mark only successful sends, formatting

- Fix WhatsApp templateComponents parameters: was array-of-arrays (one array
  per user in a batch), now sends one request per user with a flat parameters
  array as the WhatsApp API expects
- Remove unused userIds variable (no longer needed with per-user loop)
- Fix 24h notification window: replace getBrazilTomorrowEnd() (fixed calendar
  day) with a rolling window of now + 24h so the cron schedule time doesn't
  affect which events qualify
- Fix markSubscriptions: only mark a subscription as sent when the API call
  for that specific user succeeded (not all indiscriminately)
- Run prettier --write to fix formatting errors across the project
